Client: cleanup Demuxer wrapper

kprotty opened the pull request:
The vsr.Client wrapper simply took/gave bytes over types. So instead the StateMachine's Demuxer API is changed and vsr.Client simply re-exports it. 

Additionally, Demuxer takes a mutable slice to fixup `Create{T}Result.index` but vsr.Client returned an immutable slice. So Client was also changed to return a mutable slice instead of Demuxer having to `@constCast`.

kprotty committed:
client: remove vsr.Client Demuxer wrapper

The vsr.Client wrapper was only taking/giving bytes over types
so instead the StateMachine's Demuxer API is changed and
vsr.Client simply re-exports it.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -21,45 +21,19 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
                 user_data: u128,
                 operation: StateMachine.Operation,
-                results: []const u8,
+                results: []u8,
```
In `vsr.Client.on_reply`, can we assert `reply.references == 1` just before 

```
                // The message is the result of raw_request(), so invoke the user callback.
                inflight.callback(
                    inflight.user_data,
                    inflight_vsr_operation.cast(StateMachine),
                    reply.body(),
                );
```

? I'm not sure if it will work... there might be an extra reference in the message bus, but it is worth a try.

And either way, I would add a comment to that callback invocation warning that the callback is allowed to mutate the reply body in-place.

kprotty replied:
To guarantee no other references, should vsr copy the reply body to its own internal buffer and pass that to the callback?

sentientwaffle replied:
No, I was just wondering if we already had a guarantee of no other references. I don't think it is worth the overhead of copying it.

kprotty committed:
client: annotate message body lifetime

kprotty committed:
client: remove message.ref assertion

The MessageBus will try to reuse an existing reserved Message when parsing.
It does this by calling `.ref()` on it and storing it in the Connection. This seems
to make the `.references` count on the Message spuriously >1 sometimes.
